src/server/twemcache/data/process.c

```diff
@@ -198,21 +265,31 @@ _process_set(struct response *rsp, struct request *req)
 static void
 _process_add(struct response *rsp, struct request *req)
 {
-    item_rstatus_t status;
-    struct bstring *key;
+    put_rstatus_t status;
+    item_rstatus_t istatus;
+    struct item *it;
+    struct bstring key;
+
+    status = _put(&istatus, req);
+    if (status == PUT_PARTIAL) {
+        return;
+    }
 
     INCR(process_metrics, add);
-    key = array_first(req->keys);
-    if (item_get(key) != NULL) {
+    it = (struct item *)req->reserved;
+    key = (struct bstring){it->klen, item_key(it)};
```

Comment:
if `item_reserve` failed in `_put`, `req->reserved` will be null, and all this code will fail; check for `status != PUT_ERROR` earlier